examples/mcts.dx

```diff
@@ -0,0 +1,323 @@
+
+    -- write to the tree
+    new_node = get (snd_ref tree_ref)
+    base_children_ix = for i:(Fin n). new_node +| (1 + ordinal i)
+    -- update base's children list
+    new_base_node = Branch base_stats base_parent (to_list base_children_ix)
+    (fst_ref tree_ref)!node := Just new_base_node
+    -- insert every child
+    for i.
+      child_ix = base_children_ix.i
+      (fst_ref tree_ref)!child_ix := Just child_nodes.i
+    snd_ref tree_ref := new_node +| n
+
+' Simulate the game starting at a given state until the game terminates and
+return the simulation result.
+
+def mcts_rollout {move state}
+    (game: Game state move)
+    (key: Key) (curr_state: state) : Bool =
+    -- output is w.r.t. current player
+    with_state curr_state \ref. iter \i.
+      curr_state' = get ref
+      if (get_at #game_over game) curr_state'
+        then
+          -- #win_or_lose is w.r.t. the last player
+          result = (get_at #win_or_lose game) curr_state'
+          Done $ xor (is_even i) result
+        else
+          (AsList _ new_moves) = (get_at #possible_moves game) curr_state'
+          -- rollout policy: uniform random
+          new_move = new_moves.(rand_idx (hash key i))
+          ref := (get_at #update game) curr_state' new_move
+          Continue
+
+' Update the record of a node and its chain of ancestors,
+based on the simulation result.
+
+def mcts_backpropagate {move state node_ix h} [Eq node_ix] 
+    (game: Game state move)
+    (tree_ref: Ref h (Tree node_ix (NodeStats state)))
+    (node: node_ix) (newResult: Float) : {State h} Unit =
+    -- newResult: win or lose w.r.t curr_player @node
+    leaf_min_player = get (fst_ref tree_ref)!node |>
+      from_just |> node_level |> is_odd
+    maxResult = if leaf_min_player 
+                    then (1. - newResult)
+                    else newResult
+
+    -- collect itself and all ancestors
+    ancestors_list = yield_accum (ListMonoid node_ix) \ls.
+      with_state node \node_ref. iter \_.
+        append ls (get node_ref)
+        curr_node = from_just $ get (fst_ref tree_ref)!(get node_ref)
+        (Branch _ curr_parent _) = curr_node
+        if is_nothing curr_parent || (0@node_ix) == (get node_ref)
+          then Done ()
+          else
+            node_ref := from_just curr_parent
+            Continue
+
+    -- update win and loss counts
+    (AsList _ ancestors) = ancestors_list
+    for i.
+      -- unpack
+      ancestor = from_just $ get (fst_ref tree_ref)!(ancestors.i)
+      (Branch stats parent children) = ancestor
+      (wins, visits, level, state) = stats
+      -- compute
+      (new_wins, new_visits) = (wins + maxResult, visits + 1)
+      -- pack
+      new_stats = (new_wins, new_visits, level, state)
+      (fst_ref tree_ref)!(ancestors.i) := Just (Branch new_stats parent children)
+    ()
+
+' Wrap up the procedure for taking one sample and extend the tree.
+
+-- slow: every sample makes two copies of the tree array
```

Comment:
It should be possible to speed this up by making mcts_select operate on a Ref Tree instead of a Tree, and also this function itself, thus avoiding all the copies.  You could then have a pure wrapper on the outside that amortizes the tree copy over multiple mcts samples, if desired.  Or you could leave it the way it is.

We should probably have some operation that allows temporarily freezing a reference so that a pure function may be called on its contents without incurring the copy, but we don't have that yet.

Reply:
Ack. That is precisely what I would like to have one day -- an immutable/read-only reference.